Watch rewrite

eatkins opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines

eatkins committed:
Add fileInputs and watchTriggers task

This adds two new tasks: fileInputs and watchTriggers, that will be used by sbt
both to fetch files within a task as well as to create watch sources for
continuous builds. In a subsequent commit, I will add a task for a
command that will traverse the task dependency graph to find all of the
input task dependency scopes. The idea is to make it possible to easily
and accurately specify the watch sources for a task. For example, we'd
be able to write something like:

val foo = taskKey[Unit]("print text file contents")
foo / fileInputs := baseDirectory.value  **  "*.txt"
foo := {
  (foo / fileInputs).value.all.foreach(f =>
    println(s"$f:\n${new String(java.nio.Files.readAllBytes(f.toPath))}"))
}

If the user then runs `~foo`, then the task should trigger if the user
modifies any file with the "txt" extension in the project directory.
Today, the user would have to do something like:

val fooInputs = settingKey[Seq[Source]]("the input files for foo")
fooInputs := baseDirectory.value ** "*.txt"

val foo = taskKey[Unit]("print text file contents")
foo := {
  fooInputs.value.all.foreach(f =>
    println(s"$f:\n${new String(java.nio.Files.readAllBytes(f.toPath))}"))
}
watchSources ++= fooInputs.value

or even worse:

val foo = taskKey[Unit]("print text file contents")
foo := {
  (baseDirectory.value ** "*.txt").all.foreach(f =>
    println(s"$f:\n${new String(java.nio.Files.readAllBytes(f.toPath))}"))
}
watchSources ++= baseDirectory.value ** "*.txt"

which makes it possible for the watchSources and the task sources to get
out of sync.

For consistency, I also renamed the `outputs` key `fileOutputs`.

eatkins committed:
Add InputGraph

This commit adds functionality to traverse the settings graph to find
all of the Inputs settings values for the transitive dependencies of the
task. We can use this to build up the list of globs that we must watch
when we are in a continuous build. Because the Inputs key is a setting,
it is actually quite fast to fetch all the values once the compiled map
is generated (O(2ms) in the scripted tests, though I did find that it
took O(20ms) to generate the compiled map).

One complicating factor is that dynamic tasks do not track any of
their dynamic dependencies. To work around this, I added the
transitiveDependencies key. If one does something like:
foo := {
  val _ = bar / transitiveDependencies
  val _ = baz / transitiveDependencies
  if (System.getProperty("some.prop", "false") == "true") Def.task(bar.value)
  else Def.task(baz.value)
}
then (foo / transitiveDependencies).value will return all of the inputs
and triggers for bar and baz as well as for foo.

To implement transitiveDependencies, I did something fairly similar to
streams where if the setting is referenced, I add a default
implementation. If the default implementation is not present, I fall
back on trying to extract the key from the commandLine. This allows the
user to run `show bar / transitiveDependencies` from the command line
even if `bar / transitiveDependencies` is not defined in the project.

It might be possible to coax transitiveDependencies into a setting, but
then it would have to be eagerly evaluated at project definition time
which might increase start up time too much.  Alternatively, we could
just define this task for every task in the build, but I'm not sure how
expensive that would be. At any rate, it should be straightforward to
make that change without breaking binary compatibility if need be. This
is something to possibly explore before the 1.3 release if there is any
spare time (unlikely).

eatkins committed:
Add internalDependencyConfigurations setting

In order to walk the full dependency graph of a task, we need to know
the internal class path dependency configurations. Suppose that we have
projects a and b where b depends on *->compile in a. If we want to find
all of the inputs for b, then if we find that there is a dependency on
b/Compile/internalDependencyClasspath, then we must add a / Compile /
internalDependencyClasspath to the list of dependencies for the task.

I copied the setup of one of the other scripted tests that was
introduced to test the track-internal-dependencies feature to write a
basic scripted test for this new key and implementation.

eatkins commented:
This is built on top of #4476. Once that's merged in, I'll rebase this PR. While this is a fat change on it's own, it's bloated further by the ~1000 line diff due to #4476.

eatkins commented:
Fixes #3718

eatkins commented:
A few highlights:
1) There is now an alternative command `watch` that works the same as `~` but is implemented as an InputTask. This allows plugin authors to add watch functionality into a custom task. It also may be more clear for new users than `~`.
2) You can terminate a watch and return to the shell by typing 's'. This is useful if you ran, say, `sbt ~compile` but realized that you wanted to do something else. Before you'd have had to restart sbt entirely. Now you just easily drop back into an sbt shell.
3) Watch settings can be scoped at task level:
```
sbt:swoval> set Global / watchLogLevel := Level.Debug
[info] Defining Global / watchLogLevel
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to swoval (in build file:/Users/ethanatkins/work/swoval/files/)
sbt:swoval> watch filesJVM/compile
[info] [watch] [debug] meta-build - Received Update(Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Deadline(645137216846528 nanoseconds))
[info] Formatting 63 sources in file-tree-views using JavaFormatter.
[info] Formatting 6 sources in file-tree-views using ClangFormatter.
[success] Total time: 1 s, completed Jan 11, 2019 2:11:24 PM
[info] [watch] 1. Waiting for source changes in project filesJVM... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[info] [watch] [debug] meta-build - Received event for path /Users/ethanatkins/work/swoval/files/.gitignore
[info] [watch] [debug] Received input events: CancelWatch. Taking CancelWatch
[success] Total time: 2 s, completed Jan 11, 2019 2:11:26 PM
sbt:swoval> set filesJVM / watchLogLevel := Level.Info
[info] Defining filesJVM / watchLogLevel
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to swoval (in build file:/Users/ethanatkins/work/swoval/files/)
sbt:swoval> watch filesJVM/compile
[info] Formatting 63 sources in file-tree-views using JavaFormatter.
[info] Formatting 6 sources in file-tree-views using ClangFormatter.
[success] Total time: 0 s, completed Jan 11, 2019 2:11:48 PM
[info] [watch] 1. Waiting for source changes in project filesJVM... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[success] Total time: 4 s, completed Jan 11, 2019 2:11:50 PM
sbt:swoval> watch filesJS/compile
[info] [watch] [debug] meta-build - Received Update(Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Entry(/Users/ethanatkins/work/swoval/files/.gitignore, Right(TypedPath(/Users/ethanatkins/work/swoval/files/.gitignore))),Deadline(645167057517128 nanoseconds))
[info] stderr: info TOOL Using Ninja generator, because ninja is available.
[info] info CMD BUILD
[info] info RUN cmake --build "/Users/ethanatkins/work/swoval/files/files/js/npm/build" --config Release
[info]
[info] stdout: ninja: no work to do.
[info]
[success] Total time: 2 s, completed Jan 11, 2019 2:11:56 PM
[info] [watch] 1. Waiting for source changes in project filesJS... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[info] [watch] [debug] meta-build - Received event for path /Users/ethanatkins/work/swoval/files/.gitignore
[info] [watch] [debug] Received input events: CancelWatch. Taking CancelWatch
[success] Total time: 9 s, completed Jan 11, 2019 2:12:02 PM
sbt:swoval> set Global / watchLogLevel := Level.Info
[info] Defining Global / watchLogLevel
[info] The new value will be used by no settings or tasks.
[info] Reapplying settings...
[info] Set current project to swoval (in build file:/Users/ethanatkins/work/swoval/files/)
sbt:swoval> watch filesJS/compile
[success] Total time: 0 s, completed Jan 11, 2019 2:12:14 PM
[info] [watch] 1. Waiting for source changes in project filesJS... (press enter to interrupt, 'r' to re-run the command or 's' to return to the shell)
[success] Total time: 5 s, completed Jan 11, 2019 2:12:18 PM
```
4) @japgolly There is now a watchOnEnter setting that can run arbitrary code before any of the watch processing begins. This can be used to clear the console.

eatkins commented:
@OlegYch this PR resolves #3718 by introducing a new data structure `Glob` that looks like
```scala
sealed trait Glob {
  def base: File
  def filter: FileFilter
  def depth: Int
}
```
I added two setting keys, `inputs` and `triggers` that each have return type Seq[Glob]. Suppose I was adding a protoc command, then I could do it with:
```scala
val protoc = TaskKey[Unit]("compile protobuf files")
protoc / inputs += baseDirectory.value / "src" / "main" / "proto" ** "*.proto"
def protocImpl(files: Seq[File]): Unit = ???
protoc := {
  val sourceFiles = (protoc / inputs).value.all
  protocImpl(sourceFiles)
}
```
With these definitions `~protoc` will automatically track all of the "*.proto" files described by the `protoc / inputs` definition.
Internally this is how compile now works after these changes.

If you wanted to make protoc trigger automatically if a non-dependent file changes, that can be done with e.g.:
```scala
protoc / triggers += baseDirectory.value / "recompile-protobufs.txt"
```
This way the inputs and triggers are handled separately so that the protocImpl doesn't get the "recompile-protobufs.txt" file, but that file will still trigger a new build.

eatkins commented:
fixes #4445

eatkins committed:
Add StateTransform class

This commit cleans up the approach for transforming the sbt state upon
completion of a task returning State. I add a new approach where a task
can return an instance of StateTransform, which is just a wrapper around
State. I then update EvaluateTask to apply this stateTransform rather
than the (optional) state transformation that may be stored in the Task
info parameter. By requiring that the user return StateTransform rather
than State directly, we ensure that no existing tasks that depend on the
state transformation function embedded in the Task info break. In sbt 2,
I could see the possibility of making this automatic (and probably
removing the state transformation function via attribute).

The problem with using the transformState attribute key is that it is
applied non-deterministically. This means that if you decorate a task
returning State, then the state transformation may or may not be
correctly applied.

I tracked this non-determinism down to the stateTransform
method in EvaluateTask. It iterates through the task result map and
chains all of the defined transformState attribute values. Because the
result is a map, this order is not specified. This chaining is arguably
a bad design because State => State does not imply commutivity. Indeed,
the problem here was that my state transformation functions were
constant functions, which are obviously non-commutative. I believe that
this logic likely written under the assumption that there would be no
more than one of these tranformations in a given result map.

OlegYch commented:
@eatkins this looks great to me
is there a plan of migrating core tasks to use it?

eatkins commented:
@OlegYch I'm not sure how many core tasks really need to be migrated. I've already made compile work with the new infrastructure. Thus any task that depends on compile transitively works with the new watch infrastructure. I'm actually not aware of many built in tasks that depend on a set of source files. Any tasks that do, however, should be fairly straightforward to migrate.

Not directly related to watch, but I also use globs to improve io performance. I tacked on 
https://github.com/eatkins/sbt/commit/8983e0983b9c1201cb04556565af0bba422fce72 to this PR to update the clean task to use globs. The time to evaluate clean in the sbt project dropped from 6s to < 1s on my machine.

eatkins committed:
Support inputs in dynamic tasks

Prior to this commit, it was necessary to add breadcrumbs for every
input that is used within a dynamic task. In this commit, I rework the
watch setup so that we can track the dynamic inputs that are used. To
simplify the discussion, I'm going to ignore aggregation and
multi-commands, but they are both supported. To implement this change, I
update the GlobLister.all method to take a second implicit argument:
DynamicInputs. This is effectively a mutable Set of Globs that is
updated every time a task looks up files from a glob. The repository.get
method should already register the glob with the repository. The set of
globs are necessary because the repository may not do any file filtering
so the file event monitor needs to check the input globs to ensure that
the file event is for a file that actually requested by a task during
evaluation.

* Long term, I plan to add support for lifting tasks into a dynamic task
in a way that records _all_ of the possible dependencies for the task
through each of the dynamic code paths. We should revisit this change to
determine if its still necessary after that change.

eatkins committed:
Refactor Watched

This is a huge refactor of Watched. I produced this through multiple
rewrite iterations and it was too difficult to separate all of the
changes into small individual commits so I, unfortunately, had to make a
massive commit. In general, I have tried to document the source code
extensively both to facilitate reading this commit and to help with
future maintenance.

These changes are quite complicated because they provided a built-in
like api to a feature that is implemented like a plugin. In particular,
we have to manually do a lot of parsing as well as roll our own
task/setting evaluation because we cannot infer the watch settings at
project build time because we do not know a priori what commands the
user may watch in a given session. The dynamic setting and task
evaluation is mostly confined to the WatchSettings class in Continuous.
It feels dirty to do all of this extraction by hand, but it does seem to
work correctly with scopes.

At a high level this commit does four things:
1) migrate the watch implementation to using the InputGraph to collect
   the globs that it needs to monitor during the watch
2) simplify WatchConfig to make it easier for plugin authors to write
   their own custom watch implementations
3) allow configuration of the watch settings based on the task(s) that
   is/are being run
4) adds an InputTask implemenation of watch.

Point #1 is mostly handled by Point #3 since I had to overhaul how _all_
of the watch settings are generated. InputGraph already handles both
transitive inputs and triggers as well as legacy watchSources so not
much additional logic is needed beyond passing the correct scoped keys
into InputGraph.

Point #3 require some structural changes. The watch settings cannot in
general be defined statically because we don't know a priori what tasks
the user will try and watch. To address this, I added code that will
extract the task keys for all of the commands that we are running. I
then manually extract the relevant settings for each command. Finally, I
aggregate those settings into a single WatchConfig that can be used to
actually implement the watch. The aggregation is generally
straightforward: we run all of the callbacks for each task and choose
the next watch state based on the highest priority Action that is
returned by any of the callbacks.

Because I needed Extracted to pull out the necessary settings, I was
forced to move a lot of logic out of Watched and into a new singleton,
Continuous, that exists in the main project (Watched is in the command
project). The public footprint of Continuous is tiny. Even though I want
to make the watch feature flexible for plugin authors, the
implementation and api remain a moving target so I do not want to be
limited by future binary compatibility requirements. Anyone who wants to
live dangerously can access the private[sbt] apis via reflection or by
adding custom code to the sbt package in their plugin (a technique I've
used in CloseWatch).

Point #2 is addressed by removing the count and lastStatus from the
WatchConfig callbacks. While these parameters can be useful, they are
not necessary to implement the semantics of a watch. Moreover, a status
boolean isn't really that useful and the sbt task engine makes it very
difficult to actually extract the previous result of the tasks that were
run. After this refactor, WatchConfig has a simpler api. There are fewer
callbacks to implement and the signatures are simpler. To preserve the
_functionality_ of making the count accessible to the user specifiable
callbacks, I still provided settings like watchOnInputEvent that accept
a count parameter, but the count is actually tracked externally to
Watched.watch and incremented every time the task is run.

Moreover, there are a few parameters of the watch: the logger and
transitive globs, that cannot be provided via settings. I provide
callback settings like watchOnStart that mirror the WatchConfig
callbacks except that they return a function from Continuous.Arguments
to the needed callback. The Continuous.aggregate function will check if
the watchOnStart setting is set and if it is, will pass in the needed
arguments. Otherwise it will use the default watchOnStart implementation
which simulates the existing behavior by tracking the iteration count in
an AtomicInteger and passing the current count into the user provided
callback. In this way, we are able to provide a number of apis to the
watch process while preserving the default behavior.

To implement #4, I had to change the label of the `watch` attribute key
from "watch" to "watched". This allows `watch compile` to work at the
sbt command line even thought it maps to the watchTasks key. The actual
implementation is almost trivial. The difference between an
InputTask[Unit] and a command is very small. The tricky part is that the
actual implementation requires applying mapTask to a delegate task that
overrides the Task's info.postTransform value (which is used to
transform the state after task evaluation). The actual postTransform
function can be shared by the continuous task and continuous command.
There is just a slightly different mechanism for getting to the state
transformation function.

eatkins commented:
I dropped the WIP from this PR. It still depends on #4539 (and its dependent PRs), but I consider it ready to go modulo code review feedback.

eatkins committed:
Remove unneeded string interpolation

eatkins committed:
Create Watch.scala

I decided that it makes sense to move all of the new watch code out of
the Watched companion object since the Watched trait itself is now
deprecated. I don't really like having the new code in Watched.scala
mixed with the legacy code, so I pulled it all out and moved it into the
Watch object. Since we have to put all of the logic for the Continuous
object in main in order to access the sbt.Keys object, it makes sense to
move the logic out of main-command and into command so that most of the
watch related logic is in the same subproject.

eatkins committed:
Improve watch messages

This commit reworks the watch start message so that instead of printing
something like:

[info] [watch] 1. Waiting for source changes... (press 'r' to re-run the command, 'x' to exit sbt or 'enter' to return to the shell)

it instead prints something like:

[info] 1. Monitoring source files for updates...
[info] Project: filesJVM
[info] Command: compile
[info] Options:
[info]   <enter>: return to the shell
[info]   'r': repeat the current command
[info]   'x': exit sbt

It will also print which path triggered the build.

eatkins committed:
Clean up file repository management

I had needed to add proxy classes for the global FileTreeRepository so
that tasks that called the close method wouldn't actually stop the
monitoring done by the global repository. I realized that it makes a lot
more sense to just not provide direct access to the underlying file tree
repository and let the registerGlobalCaches manage its life cycle
instead.

eatkins commented:
@eed3si9n I consider this functionally complete. There are still some api changes I am making to glob and io that I will handle in a separate PR, but this would be usable for an M3.

eed3si9n reviewed (commented):
The commit message on e2c971460a088b2c9e44500ac6bd5ded71194476 seems a bit outdated.

> One complicating factor is that dynamic tasks do not track any of
their dynamic dependencies. To work around this, I added the
transitiveDependencies key.

eed3si9n commented on main/src/main/scala/sbt/Keys.scala:
```diff
@@ -133,6 +149,8 @@ object Keys {
   val sources = taskKey[Seq[File]]("All sources, both managed and unmanaged.").withRank(BTask)
   val sourcesInBase = settingKey[Boolean]("If true, sources from the project's base directory are included as main sources.")
+  val inputs = settingKey[Seq[Glob]]("The file globs that are used by a task. This setting will generally be scoped per task. It will also be used to determine the sources to watch during continuous execution.")
+  val triggers = settingKey[Seq[Glob]]("Describes files that should trigger a new continuous build.")
```
1. When would you need this, instead of `foo / inputs`?
2. `watchTriggers` might be a better name since the trigger might come up in different context like `def trigger = allRequirements`

eatkins replied:
Yeah. That's a good point. I also am planning to move these keys into the plugin we discussed.

eed3si9n commented on main/src/main/scala/sbt/internal/FileManagement.scala:
```diff
@@ -55,11 +53,8 @@ private[sbt] object FileManagement {
-        val service = Watched.createWatchService(pollInterval)
-        FileTreeRepository.legacy(FileAttributes.default _, (_: Any) => {}, service)
-      }
+      val service = Watched.createWatchService(pollInterval)
```
Is this undoing Windows specific handling added in https://github.com/sbt/sbt/commit/e3625ca8b0781288f379eff6d6e6534998d64200, or it's no longer needed?

eatkins replied:
It's no longer needed.

eatkins replied:
Actually, that is mistaken. It is not needed using new code that isn't part of this PR so I need to restore the windows specific code until I open that PR. It works fine on my VM, but CI fails with the regular file repository. I'm 90% sure that CI will pass with some changes from https://github.com/sbt/io/pull/226.

eed3si9n commented:
The commit message on d328d389cc7fd39eb75e4ea819a4e7721be7fc2d looks unfinished.

eatkins commented:
Drat. I just independently noticed that I didn't finish the commit message to https://github.com/sbt/sbt/commit/d328d389cc7fd39eb75e4ea819a4e7721be7fc2d. I'm fixing the scripted tests now and will fix that.

eatkins commented:
> The commit message on [e2c9714](https://github.com/sbt/sbt/commit/e2c971460a088b2c9e44500ac6bd5ded71194476) seems a bit outdated.
> 
> > One complicating factor is that dynamic tasks do not track any of
> > their dynamic dependencies. To work around this, I added the
> > transitiveDependencies key.

Yeah. I should change it to say that that issue is addressed in subsequent commits.

eatkins committed:
Validate the cache by default

This commit change the default FileTree.Repository to always use a polling file
repository but one that validates the current file system results
against the cache results. On windows, we do not validate the cache
because the cache can cause io contention in scripted tests. The
cache does seem to work ok on my VM, but not on appveyor for whatever
reason. Validating the cache by default was suggested by @smarter in a
comment in https://github.com/sbt/sbt/issues/4543.

eed3si9n commented:
AppVeyor failed, but I think it's unrelated Windows flakiness, so I'm going to rerun the job.

eed3si9n commented on .appveyor.yml:
```diff
@@ -20,7 +20,7 @@ install:
       }
   - SET PATH=C:\sbt\sbt\bin;%PATH%
-  - SET SBT_OPTS=-XX:MaxPermSize=2g -Xmx4g -Dfile.encoding=UTF8
+  - SET SBT_OPTS=-XX:MaxPermSize=2g -Xmx4g -Dfile.encoding=UTF8 -Dsbt.supershell=never
```
We should split this commit out to another PR.
Maybe we should turn it off by default in CI.
https://github.com/sbt/sbt/blob/ca5f978df495a5a975302cedba73986d2abaa262/main/src/main/scala/sbt/Defaults.scala#L332

eatkins replied:
Ok. I added #4590.  I'll remove this commit from this PR.

eatkins commented:
I think the appveyor flakiness is real. It's due to io contention between the cache and the various task threads listing files and computing the file stamps. Linux and osx generally don't lock files which is why we don't see the errors on travis. I'm working on fixing it.

eatkins commented:
I think the appveyor issues should be resolved now. They passed on my local appveyor instance at least.

eed3si9n reviewed (dismissed):
LGTM

eatkins commented:
Thanks for the approval! I had to resolve a merge conflict from #4590 so I need a re-approval for this pending CI.
